### ISSUE TITLE
Editorial-only NITS on QUIC-Transport-24

### ISSUE THREAD

**Issue by gorryfair** [quicwg/base-drafts#3214]
(Edited by @larseggert  on Aug 19.)

~~/There are certain operations which an application MUST /~~
~~- ought this to be /which an /that an/~~
@larseggert says: current text has changed, @gorryfair to review

~~/However, implementations MAY choose to offer the ability to
   deliver data out of order to a receiving application./~~
~~- This doesn’t say the default is in-order delivery, which I would have thought was the case. It also doesn’t say whether this is per-connection, or per-stream (I imagined the latter).~~
#4017 

~~/frames in multiple packets in order to make/~~
~~- Could be read as implying sequencing, I suggest removing /in order/ or saying /to ensure that~~
#4018

~~/A receiver maintains a cumulative sum of
   bytes received on all streams, which is used to check for flow
   control violations./~~
~~- It would have helped me, had this explained what was intended by “control violations” - I think it means to detect when the sender sends more than permitted by the flow control limit. I think the control part comes in a later sentence? (Is this measured against the largest advertised offset, or the current offset? … given what is written about reducing the advertised value?)~~
#4019

~~/However, multiplexing
   connections on the same local IP address and port while using zero-
   length connection IDs will cause failures in the presence of peer
   connection migration, NAT rebinding, and client port reuse; and
   therefore MUST NOT be done unless an endpoint is certain that those
   protocol features are not in use./~~
~~- Seems like dodgy requirements text to me. I likely agree with the conclusion, but what is the MUST actually determining when the sentence starts with a however and includes ‘certain’ … ’not in use”? I suspect it would be better to have a full stop after the examples of failure and then state what is required.~~
@larseggert says: address in PR

~~/An endpoint MAY limit the frequency or the total
   number of connection IDs issued for each connection to avoid the risk
   of running out of connection IDs; see Section 10.4.2./~~
~~- I don’t understand /frequency/ or /total number/ in this sentence. I had expected a limit on the outstanding number of unused connections or the total number of in-use (used or available) connections, but the words I saw didn’t make sense to me.~~
@larseggert says: addressed in PR

~~/An endpoint that migrates
   away from a local address SHOULD retire all connection IDs used on
   that address once it no longer plans to use that address./~~
~~- I expected to know what happens if the should is not respected, and wondered in this case whether the resources remain committed to the previous address and would become unable to be utilised?~~
@larseggert says: I can no longer find this text in the draft, @gorryfair to check

~~/the peer MUST retire the corresponding connection IDs
   and send corresponding RETIRE_CONNECTION_ID frames.  Failing to
   retire /~~
~~- The text requires the sender to do something, ut clearly reordering and other things could follow - as it says in the next text - so why is this a MUST, not a SHOULD - with strong indication do otherwise will have implications?~~
@larseggert says: text has changed significantly in draft, @gorryfair to review

~~In section 5.2 /Hosts try to/~~
~~- Is host the same endpoints, as elsewhere, or specifically the node hosting the stack? and why?~~
@larseggert says: I can no longer find this text in the draft, @gorryfair to check

~~/Packets that don't match an existing connection are discarded ~~
~~- I’d personally really prefer to avoid using “don’t” and “isn’t” (which currently occur a few times in this section).~~
~~-  also can the ID explicitly say which field did not match the ID.~~
@larseggert says: addressed in PR

~~/Servers that do not support a particular version
   are unlikely to be able to decrypt the payload of the packet.
   Servers SHOULD NOT attempt to decode or decrypt/~~
~~- This seems like tuning? Or is the intention to protect the endpoint from an attack, please explain the SHOULD.~~
#4020

~~/   Servers MUST drop incoming packets under all other circumstances./~~
~~- Normally in other transport protocols, this requirement would be accompanied by a “and SHOULD log the event” to enable people to understand what has happened.~~
@larseggert says: I am not aware of other IETF specs that require logging.

~~/There are certain operations which an application MUST/~~
~~- I think /which/ needs to /that/.~~
@larseggert says: already addressed

~~/ The size of the first packet sent by a client will determine whether
   a server sends a Version Negotiation packet.  Clients that support
   multiple QUIC versions SHOULD pad the first packet they send to the
   largest of the minimum packet sizes across all versions they support.
   This ensures that the server responds if there is a mutually
   supported version.
/~~
~~- Are you saying a client that does not PAD does not negotiate? If so: What is the logic behind that design decision?~~
~~- I don’t read that as it assures what it say. To me, the padding assures the client can send this side of packet to the remote endpoint.~~
~~- Does this agree with what is written in section 8.1 on padding? and Section 14?~~
#4021

~~/An endpoint can verify support for Explicit Congestion Notification
   (ECN) in the first packets it sends, /~~
~~- Nearly - It can verify the endpoints support ECN and the current path could be used to transfer ECN-marked packets. That’s not quite enough to complete verification that the path and endpoints have full ECN support.~~
#4022

~~/MUST specify whether they MUST,/~~
~~- could be written is /is REQUIRED to specify whether they MUST/~~
@larseggert: see PR

~~/ and absent new values from the server, the default value./~~
~~- what does ‘absent’ mean in this clause, could you reword?~~
#4023

~~/An endpoint MUST ignore transport parameters that it does
   not support./~~
~~- What happens if the parameter definition requires that it is stored? How would it know?~~
#4024

~~/Address validation is used by QUIC to avoid being used for a traffic
   amplification attack.  /~~
~~- This doesn’t parse to me, /is used…being bused/, please could you rephrase?~~
@larseggert says: I can no longer find this text in the draft, @gorryfair to check

~~/A token SHOULD be constructed in a way that allows the server to
   distinguish it from tokens that are sent in Retry packets as they are
   carried in the same field./~~
~~- I’m not sure whether this intended /as/ or /because/ - but either way I did not understand that sentence.~~
@larseggert says: current text has changed, @gorryfair to review

~~/claimed client address (IP
   and port)/~~
~~- What is IP and port? - is this both IP addresses or the remote address? (do you care about Address Family?) and pair of transport ports (or one of the ports)~~
@larseggert says: I can no longer find this text in the draft, @gorryfair to check

~~/MUST respond
   immediately by echoing the data/~~
~~- Of course it may not be able to immediately send anything:-) , but I suspect this needs to not delay?~~
#4025

~~/as it might be a result of a forwarded
   packet (see Section 9.3.3) or misrouting. /~~
~~- I think “forwarding” in a transport spec means normal network behaviour, and routing is the thing controls that. nIf the spec needs to use these words can they be prefixed with something “proxy forwarding - or application forwarding”- or whatever to indicate that these are middleboxes, rather than routers/switches.~~
@larseggert says: I can no longer find this text in the draft, @gorryfair to check

~~/9.2.  Initiating Connection Migration/~~
~~- I assert the new path may not have the same RTT, because the path could traverse a different set of network devices or be subject to different policies that influence forwarding behaviour. These are not currently listed here, but are the basis for some later text in 9.3.1. Can we mention here?~~
#4026

~~/This style of attack relies on the attacker using a path that is
   approximately as fast/~~
~~- In the sense of this phrase, I think “fast” is not referring to line rate or transfer speed, but to the same amount of time … I think the word needs to be replaced.~~
@larseggert says: see PR

~~/   At any time, endpoints MAY change the Destination Connection ID they
   send to a value that has not been used on another path./~~
~~- I understand the idea of changing, but what is the “send to a value” text about?~~
@larseggert says: see PR

~~/If path validation succeeds, the client SHOULD immediately begin
   sending all future packets to the new server address/~~
~~- /immediately begin sending/ troubles me here, and I would prefer this to also explicitly say that is subject to 9.4 (because we all know people read documents with certain emphasis, and I wouldn’t like this specific detail to be missed.~~
@larseggert says: see PR

~~/Migrating a connection to a new server address mid-connection is left
   for future work./~~
~~- I’d prefer to say is not described in this spec or is not supported by this spec - rather than these words, I think the SHOULD indicates it is likely the former.~~
@larseggert says: see PR

~~Section 11:
/   An endpoint that detects an error SHOULD signal the existence of that
   error to its peer. /~~
~~- Normally in other transport protocols, this requirement would be accompanied by a “and SHOULD log the event” to enable people to understand what has happened.~~
@larseggert says: I am not aware of other IETF specs that require logging.

~~This statement is dangerous:~~
~~/However, an implementor should
   only deviate from these requirements after careful consideration of
   the performance implications of doing so./~~
~~- Let me argue this statement needs to be replaced by:~~
~~- “The requirements described are baed on consideration of~~
~~   the performance implications and the on other flows that could share network capacity”.~~
~~- I strongly argue that implementors how do not implement a PS do not conform to that PS. That happens, but I don’t see the point in a spec saying if you don’t comply you’re still compliant.~~
#4027

~~/Limiting ACK frames avoids an infinite feedback loop of acknowledgements/~~
~~- Unclear, is the text about limiting which sets of frames result in ACK frames? or something else?~~
@larseggert says: current text has changed, @gorryfair to review

~~In Section 13.4 ~~
~~/An endpoint
   validates the use of ECN on the path, both during connection
   establishment and when migrating to a new path (Section 9)./~~
~~- I think the QUIC endpoint MAY also validate ECN is still support by the path during communication. This may be necessary when ECN is enabled, but a path change results in there no longer being support.~~
~~- I think this needs to also be permitted in Section  13.4.2, allowing QUIC to suspend ECN usage for a path or to an endpoint for which it currently has a connection.~~
#4028

~~/Note that this requires being able
   to read the ECN codepoints from the enclosing IP packet, which is not
   possible on all platforms./~~
~~- That sounds like it could be an issue, which it is not. Could we add “On those platforms, QUIC will not enable ECN for this endpoint”.~~
@larseggert says: see PR, propose to remove the sentence

~~/To provide robust connectivity in
   the presence of such devices, each endpoint independently validates
   ECN counts and disables ECN if errors are detected./~~
~~- specifically, I think this could be: /To provide robust connectivity in
   the presence of such devices, each endpoint independently validates
   ECN counts and disables ECN if the path is does not show consistent support for ECN./~~
@larseggert says: see PR

~~/   Even if validation fails, an endpoint MAY revalidate ECN on the same
   path at any later time in the connection./~~
~~- That’s possible, meaning try again to use ECN? I think the text needs to be clearer, that this can be used to enable ECN to be used within the connection.~~
#4029

~~/path supports a reasonable Maximum Transmission Unit (MTU)./~~
~~- I think this should read /path supports a reasonable Path Maximum Transmission Unit (PMTU)/~~
@larseggert says: already addressed

~~Also:~~
~~/   Prior to validating the client address, servers MUST NOT send more
   than three times as many bytes as the number of bytes they have
   received./~~
~~- So, let me understand - does that include retransmissions? or is this constrain within the PTO or something?~~
#4030

**Comment by dtikhonov** [comment]
Each of these two dozen or so nits requires its own GitHub issue.  It will become unwieldy quickly were we to start discussing all of them in one place.  This is likely there have been no comments since this issue was opened.

**Comment by gorryfair** [comment]
I'll let the editors propose how to handle these. If you want to make each of these an issue we can do that... or I can submit them as WGLC comments or we can just discuss using the list.- I still expect few (if any) of these are likely to result in important technical changes.

**Comment by dtikhonov** [comment]
I, too, think that editors should edit.  I take back my "requires its own issue" remark -- I did not mean to demand that it be done.  On the other hand, since the issue has not been commented upon, I presume that the editors' expectation is that PRs be opened.

**Comment by ianswett** [comment]
I expect once most of the design issues have been closed, the editors will focus more on closing out remaining editorial issues.  In the meantime, if there are small editorial issues I know how to fix, I try to fix them, but it's best effort at the moment.

As always, PRs to fix editorial issues are appreciated.

**Comment by MikeBishop** [comment]
PRs are a great way to fix *simple* editorial issues -- spelling, broken references, etc.  Things get dicey when an "editorial" PR changes the meaning or flow of the text and we don't want to take the whole of a given PR.  Then an editor needs to either accept the PR and partially revert it, or they need to duplicate the work for the things they want to keep.

Without looking super closely at the list of changes here, I probably wouldn't want to get a single PR that made all of these changes at once.  (Noting, of course, that this isn't my document.)

So my personal recommendation:
- Combination PRs for totally obvious editorial fixes
- Separate PRs for more invasive or controversial editorial fixes
- Issues for design changes and/or editorial issues where the solution isn't obvious (or you don't want to write it yourself)

**Comment by martinthomson** [comment]
Mike's advice is almost perfect, though I would point out that an issue does give us more opportunity to discuss things in a place that isn't a PR, so issues for the category of more intricate editorial changes can be useful.

**Comment by larseggert** [comment]
I went through @gorryfair's original list of things, and crossed those out that I think are overtaken by edits, or for which I have a proposal in #4016. There are still others in the list that need someone to take a look.

**Comment by larseggert** [comment]
OK, I *think* this is all the ones that #4016 didn't cover yet. I will leave this open to discuss the resolution in #4016, but please also look at the new tickets.
